tx/textual/valuerenderer/coins.go

```diff
@@ -0,0 +1,171 @@
+package valuerenderer
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	bankv1beta1 "cosmossdk.io/api/cosmos/bank/v1beta1"
+	basev1beta1 "cosmossdk.io/api/cosmos/base/v1beta1"
+	"cosmossdk.io/math"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+// NewCoinsValueRenderer returns a ValueRenderer for SDK Coin and Coins.
+func NewCoinsValueRenderer(q CoinMetadataQueryFn) ValueRenderer {
+	return coinsValueRenderer{q}
+}
+
+type coinsValueRenderer struct {
+	// coinMetadataQuerier defines a function to query the coin metadata from
+	// state.
+	coinMetadataQuerier CoinMetadataQueryFn
+}
+
+var _ ValueRenderer = coinsValueRenderer{}
+
+func (vr coinsValueRenderer) Format(ctx context.Context, v protoreflect.Value, w io.Writer) error {
+	if vr.coinMetadataQuerier == nil {
+		return fmt.Errorf("expected non-nil coin metadata querier")
+	}
+
+	// Check whether we have a Coin or some Coins.
+	switch protoCoins := v.Interface().(type) {
+	// If it's a repeated Coin:
+	case protoreflect.List:
+		{
+			coins, metadatas := make([]*basev1beta1.Coin, protoCoins.Len()), make([]*bankv1beta1.Metadata, protoCoins.Len())
+			var err error
+			for i := 0; i < protoCoins.Len(); i++ {
+				coin := protoCoins.Get(i).Interface().(protoreflect.Message).Interface().(*basev1beta1.Coin)
+				coins[i] = coin
+				metadatas[i], err = vr.coinMetadataQuerier(ctx, coin.Denom)
+				if err != nil {
+					return err
+				}
+			}
+
+			formatted, err := formatCoins(coins, metadatas)
+			if err != nil {
+				return err
+			}
+
+			_, err = w.Write([]byte(formatted))
+			return err
+		}
+	// If it's a single Coin:
+	case protoreflect.Message:
+		{
+			coin := v.Interface().(protoreflect.Message).Interface().(*basev1beta1.Coin)
+
+			metadata, err := vr.coinMetadataQuerier(ctx, coin.Denom)
+			if err != nil {
+				return err
+			}
+
+			formatted, err := formatCoin(coin, metadata)
+			if err != nil {
+				return err
+			}
+
+			_, err = w.Write([]byte(formatted))
+			return err
+		}
+	default:
+		return fmt.Errorf("got invalid type %t for coins", v.Interface())
+	}
+
+}
+
+func (vr coinsValueRenderer) Parse(_ context.Context, r io.Reader) (protoreflect.Value, error) {
+	panic("implement me")
+}
+
+// formatCoin formats a sdk.Coin into a value-rendered string, using the
+// given metadata about the denom. It returns the formatted coin string, the
+// display denom, and an optional error.
+func formatCoin(coin *basev1beta1.Coin, metadata *bankv1beta1.Metadata) (string, error) {
```

Comment:
The reason is #11799: the `tx` module is a standalone go module, so it cannot depend on the root module. 

I'll discuss with the SDK team on moving sdk.Coins into another module, probably `core`.

Reply:
I actually don't think it makes sense to bring in all of the existing `sdk.Coin` stuff into core. I commented in #13155. The implementation for `SIGN_MODE_TEXTUAL` will actually be slightly different from what's in `sdk.Coin`. We need bijectivity here and `ParseDecCoin` depends on the global denom regex which is actually too restrictive for a correct implementation for `SIGN_MODE_TEXTUAL` (besides being a mutable global variable). And we will need to do an extra round of allocation if we use `sdk.Coin` so that approach will be less performant (though likely not in a meaningful way I'll admit).